### PR TITLE
[feature] Show info for pending replies, allow implicit accept of pending replies

### PR DESCRIPTION
**Pull request by tsmethurst**
# Description

> If this is a code change, please include a summary of what you've coded, and link to the issue(s) it closes/implements.
>
> If this is a documentation change, please briefly describe what you've changed and why.

This pull request adds two things:

1. A little note similar to the attachment placeholder to inform a targeted user that a reply they're looking at via the API is pending their approval.
2. Implicit approval of a pending reply via liking, replying-to, or boosting the reply.

image

## Checklist

Please put an x inside each checkbox to indicate that you've read and followed it: `[ ]` -> `[x]`

If this is a documentation change, only the first checkbox must be filled (you can delete the others if you want).

- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have not leveraged AI to create the proposed changes.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [x] I/we have made any necessary changes to documentation.
- [x] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

File: internal/api/client/statuses/status_test.go [hunk 1]
```diff
@@ -59,6 +66,113 @@ type StatusStandardTestSuite struct {
 	statusModule *statuses.Module
 }
 
+// Normalizes a status response to a determinate
```
Comment (tsmethurst) [inline internal/api/client/statuses/status_test.go:69]: We could probably put this in the test rig in a later PR as it's quite useful.

File: internal/api/client/statuses/status_test.go [hunk 2]
```diff
@@ -59,6 +66,113 @@ type StatusStandardTestSuite struct {
...
+	enc.SetIndent("", "  ")
+
+	if err := enc.Encode(&rawMap); err != nil {
+		suite.FailNow(err.Error())
+	}
+
+	return strings.TrimSpace(out.String()), recorder
+}
+
+func (suite *StatusStandardTestSuite) determinateStatus(rawMap map[string]any) {
```
Comment (tsmethurst) [inline internal/api/client/statuses/status_test.go:105]: Ditto.

File: internal/typeutils/util.go [hunk 3]
```diff
@@ -187,6 +190,47 @@ func placeholderAttachments(arr []*apimodel.Attachment) (string, []*apimodel.Att
...
+		// can view + approve/reject the interaction request.
+		//
+		// Eg., https://example.org/settings/user/interaction_requests/01J5QVXCCEATJYSXM9H6MZT4JR
+		settingsURL = proto + "://" + host + "/settings/user/interaction_requests/" + intReq.ID
+	)
+
+	var note strings.Builder
+	note.WriteString(`<hr>`)
+	note.WriteString(`<p><i lang="en">ℹ️ Note from ` + host + `: `)
+	note.WriteString(`This reply to your status is pending your approval. You can accept the reply by liking, replying to, or boosting it. You can also accept or reject the reply at the following link: `)
```
Comment (NyaaaWhatsUpDoc) [inline internal/typeutils/util.go:224]: how about: `"This reply is pending your approval. You can quickly accept it by liking, boosting or replying to it. You can also accept or reject it at the following link: "`

just to reduce the number of characters (i always find shorter things easier to read, personally).

File: internal/typeutils/util.go [hunk 4]
```diff
@@ -187,6 +190,47 @@ func placeholderAttachments(arr []*apimodel.Attachment) (string, []*apimodel.Att
...
+	note.WriteString(`<a href="` + settingsURL + `" `)
+	note.WriteString(`rel="noreferrer noopener" target="_blank">`)
+	note.WriteString(settingsURL + ` (opens in a new tab)`)
```
Comment (NyaaaWhatsUpDoc) [inline internal/typeutils/util.go:227]: same here re: the "(opens in a new tab)", i think this type of warning is only useful when they *don't open in a new tab* imo.

File: internal/typeutils/internaltofrontend.go [hunk 5]
```diff
@@ -800,26 +800,55 @@ func (c *Converter) TagToAPITag(ctx context.Context, t *gtsmodel.Tag, stubHistor
...
+// to placehold unknown attachment types, and/or add a note
+// about the status being pending and requiring approval.
+func (c *Converter) statusToAPIStatus(
+	ctx context.Context,
+	status *gtsmodel.Status,
 	requestingAccount *gtsmodel.Account,
 	filterContext statusfilter.FilterContext,
 	filters []*gtsmodel.Filter,
 	mutes *usermute.CompiledUserMuteList,
+	placeholdAttachments bool,
```
Comment (NyaaaWhatsUpDoc) [inline internal/typeutils/internaltofrontend.go:846]: do we use a value other than true for placeholder attachments? i'm wondering whether it's worth just keeping the old behaviour in this case

**Comment by NyaaaWhatsUpDoc** [comment]
a few comments but otherwise looking good :)

File: internal/typeutils/util.go (hunk 3 above)
Comment (tsmethurst) [reply internal/typeutils/util.go:224]: Oh yeah that's better :)

File: internal/typeutils/util.go (hunk 4 above)
Comment (tsmethurst) [reply internal/typeutils/util.go]: Bam! https://github.com/superseriousbusiness/gotosocial/pull/3322/commits/77d637c0679eb9b1f21897de77b6f0f3c5283f80

File: internal/typeutils/util.go (hunk 3 above)
Comment (tsmethurst) [reply internal/typeutils/util.go]: Kapow! https://github.com/superseriousbusiness/gotosocial/pull/3322/commits/77d637c0679eb9b1f21897de77b6f0f3c5283f80

**Pull request merged**